### PR TITLE
feat(api): take into account air gap and other volume modifying properties when splitting volume for transfer

### PR DESCRIPTION
**Pull request by jbleon95**
# Overview

Addresses AUTH-1794.

This PR modifies volume splitting (done with `expand_for_volume_constraints_for_liquid_classes`) to take into account the air gap volume (and conditioning and disposal volume for multi-dispense distribute, more on that later) for the _**initial**_ volume splitting. This allows one to pipette the max volume for a tip even if that volume has a non-zero air gap.

For example, if one has a 50 µL pipette (or a 50 µL tip with a P1000), and they try to aspirate 50 µL with a liquid class that has an air gap of 5 µL, it will now split that volume into two 25 µL transfers. The same thing would happen if a larger value was used that would previously be split into 50 µL, a 100 µL transfer that has 5 µL air gap for the maximum would be split into three of 33.3 repeating µL instead of two 50 µL.

What will **_not_** happen is dealing with too-large air gaps _after_ we split the volume. In other words, if for some reason you have a 30 µL air gap for both 50 and 25 µL, it'll try to split into 25 µL chunks, and then fail when trying to add the air gap. While it could be possible to try and solve for a combination of volumes that would fit, this would only further complicate the liquid class transfer functions and this will cover all default and sensible cases.

For `distribute_with_liquid_class`, we are already checking if the conditioning + disposal volume + transfer volume will fit into the pipette's working volume, but the check is still included to provide an additional level of checks, the fact that a conditioning volume of non-zero negates the aspirate air gap, and to ensure that the air gap plus a disposal volume does not go over the limit either.

This PR also adds two new methods to the `LiquidHandlingPropertyByVolume` class,  a `clear_values()` the deletes all points from the curve, and a `reset_values()` that sets it back to the defaults. This was used for testing but are also useful functions to allow users to use when modifying these `byVolume` properties.

## Test Plan and Hands on Testing

Unit tests, integration tests, and snapshot tests should capture if anything in the default cases are breaking, and ensured the following protocol passes analysis and splits volume as we'd expect.

```
requirements = {
    "robotType": "Flex",
    "apiLevel": "2.24"
}

metadata = {
    "protocolName":'Liquid Class Transfer Air Gap Splitting Test',
}

def run(protocol_context):
    tiprack = protocol_context.load_labware("opentrons_flex_96_tiprack_50ul", "C2")
    trash = protocol_context.load_trash_bin('A3')
    pipette = protocol_context.load_instrument("flex_1channel_50", "left", tip_racks=[tiprack])
    nest_plate = protocol_context.load_labware("nest_96_wellplate_2ml_deep", "D1")
    arma_plate = protocol_context.load_labware("armadillo_96_wellplate_200ul_pcr_full_skirt", "D3")

    water_class = protocol_context.define_liquid_class("water")

    transfer_props = water_class.get_for(pipette, tiprack)

    transfer_props.aspirate.retract.air_gap_by_volume.clear_values()
    # This sets a flat curve of 5 regardless of volume
    transfer_props.aspirate.retract.air_gap_by_volume.set_for_volume(0.1, 5)

    pipette.transfer_with_liquid_class(
        liquid_class=water_class,
        volume=50,
        source=arma_plate.columns()[0][:2],
        dest=nest_plate.columns()[0][:2],
        new_tip="always",
        trash_location=trash,
    )

    pipette.distribute_with_liquid_class(
        liquid_class=water_class,
        volume=24,
        source=arma_plate['A1'],
        dest=nest_plate.columns()[0][:2],
        new_tip="once",
        trash_location=trash,
    )

    pipette.consolidate_with_liquid_class(
        liquid_class=water_class,
        volume=100,
        source=arma_plate.columns()[0][:2],
        dest=nest_plate['A1'],
        new_tip="once",
        trash_location=trash,
    )
```

## Changelog

- modified `expand_for_volume_constraints_for_liquid_classes` to take into account for air gap, conditioning and disposal volume
- added a `clear_values()` and a `reset_values()` function to `LiquidHandlingPropertyByVolume`

## Review requests

I'm pretty confident this covers all default cases and all sensible cases, and that any errors that we don't account for will have appropriate errors be raised, but the distribute case is especially complicated so any extra eyes/thoughts on that could be useful.

## Risk assessment

Low-medium, the default liquid classes as we have them shouldn't have any change in behavior but this does affect all three liquid class methods.

### CODE REVIEW

File: api/src/opentrons/protocols/advanced_control/transfers/common.py [hunk 1]
```diff
@@ -98,9 +100,32 @@ def expand_for_volume_constraints_for_liquid_classes(
...
+    disposal_vol: Union[LiquidHandlingPropertyByVolume, float] = 0.0,
+    conditioning_vol: Union[LiquidHandlingPropertyByVolume, float] = 0.0,
 ) -> Generator[Tuple[float, "Target"], None, None]:
     """Split a sequence of proposed transfers to keep each under the max volume, splitting larger ones equally."""
     assert max_volume > 0
     for volume, target in zip(volumes, targets):
-        for split_volume in _split_volume_equally(volume, max_volume):
+        air_gap_volume = (
+            air_gap if isinstance(air_gap, float) else air_gap.get_for_volume(volume)
+        )
```
Comment (sanni-t) [inline api/src/opentrons/protocols/advanced_control/transfers/common.py]: For distribute, airgap volume is found as `air_gap.get_for_volume(volume + disposal_volume)`. Technically, the airgap will be calculated according to volume present in the tip at each step. But the max volume that the tip will carry before adding an air gap at any point during the multi-dispense will be (aspirate_vol + disposal_vol). So we can use this when determining the volume split.

**Review by sanni-t — CHANGES_REQUESTED** [review]
I like the 'consolidation' of logic you have done here. But the order of calculation of air gap & disposal volumes will need to change.

File: api/src/opentrons/protocol_api/core/engine/instrument.py [hunk 2]
```diff
@@ -1875,12 +1875,16 @@ def _pick_up_tip() -> WellCore:
...
+            air_gap = aspirate_air_gap_by_volume.get_for_volume(next_step_volume)
             # Take air gap into account because there will be a final air gap before the dispense
-            while total_dispense_volume + next_step_volume <= max_volume:
+            while total_dispense_volume + next_step_volume <= max_volume - air_gap:
                 total_dispense_volume += next_step_volume
                 vol_aspirate_combo.append((next_step_volume, next_source))
                 try:
                     next_step_volume, next_source = next(source_per_volume_step)
+                    air_gap = aspirate_air_gap_by_volume.get_for_volume(
+                        max(next_step_volume + total_dispense_volume, max_volume)
```
Comment (sanni-t) [inline api/src/opentrons/protocol_api/core/engine/instrument.py]: I think we can remove the check for `max`. Since the volume is really not going to exceed the tip max anyway, there's no real use for it.

File: api/src/opentrons/protocol_api/core/engine/instrument.py [hunk 3]
```diff
@@ -1967,15 +1975,14 @@ def aspirate_liquid_class(
...
             aspirate_volume=volume,
-            air_gap=(
-                aspirate_props.retract.air_gap_by_volume.get_for_volume(volume)
-                if conditioning_volume is None
-                else 0
-            ),
-            disposal_volume=0,  # Disposal volume is accounted for in aspirate vol
+            air_gap=volume_for_air_gap if conditioning_volume is None else 0,
             max_volume=self.get_working_volume(),
+            current_volume=current_volume,
```
Comment (sanni-t) [inline api/src/opentrons/protocol_api/core/engine/instrument.py:1985]: Oo! Good call

**Review by sanni-t — APPROVED** [review]
Looks good to me! 
I'm going to do an edge case walkthrough for distribute but I think it'll be fine so don't want to block this PR from merging.
Thanks for making the changes!

File: api/tests/opentrons/protocols/advanced_control/transfers/test_common_functions.py [hunk 4]
```diff
@@ -92,36 +92,62 @@ def test_expand_for_volume_constraints(
...
+            0.0,
             [(50, "dest1"), (50, "dest2")],
         ),
+        (
+            [50, 50],
+            ["dest1", "dest2"],
+            50,
+            5.0,
+            [(25, "dest1"), (25, "dest1"), (25, "dest2"), (25, "dest2")],
+        ),
```
Comment (ddcc4) [inline api/tests/opentrons/protocols/advanced_control/transfers/test_common_functions.py:117]: Can we get a test where (volume + airgap) **exactly** matches the max_volume to demonstrate that we're handling the boundary condition correctly? (In complicated code, it's easy to overlook `<` vs `<=`, for example, which we'd like to catch with a test.)

E.g., max_volume = 50, air_gap = 5, and we want to transfer 45?
